### PR TITLE
Zookeeper certificate manager

### PR DESCRIPTION
**Pull request by stefan-chivu**

### CODE REVIEW

File: gateway/loaders/zookeeper/utils.go [hunk 1]
```diff
@@ -17,6 +22,55 @@ type ZookeeperClient struct {
 	conn *zk.Conn
 }
 
+func (z *ZookeeperTargetLoader) refreshZookeeperConfig(log zerolog.Logger, targetChan chan<- *connections.TargetConnectionControl, config *configuration.GatewayConfig) error {
```
Comment (petrutlucian94) [inline gateway/loaders/zookeeper/utils.go]: nit: the method name is a bit confusing. The target config is also stored in zookeeper. How about "refreshGlobalConfig"?

File: gateway/loaders/zookeeper/utils.go [hunk 2]
```diff
@@ -17,6 +22,55 @@ type ZookeeperClient struct {
...
+
+		config.ClientTLSConfig = tlsConfig
+
+		config.Log.Info().Msg("Gateway TLS Config set. Using TLS client authentication")
+
+		controlMsg := new(connections.TargetConnectionControl)
+		controlMsg.ReconnectAll = true
+		targetChan <- controlMsg
+	} else {
+		config.Log.Info().Msg("No TLS Certificates found")
```
Comment (petrutlucian94) [inline gateway/loaders/zookeeper/utils.go:68]: nit: let's log something like "No global config found", we may eventually use it for settings other than TLS certificates.

File: gateway/loaders/zookeeper/zookeeper.go [hunk 3]
```diff
@@ -163,24 +176,20 @@ func (z *ZookeeperTargetLoader) WatchConfiguration(targetChan chan<- *connection
...
-		watchPath := CleanPath(TargetPath) + CleanPath(path)
-		_, _, eventChannel, err := z.zkClient.conn.GetW(watchPath)
-		channelMap[watchPath] = eventChannel
-		if err != nil {
-			z.config.Log.Error().Err(err).Msgf("[ZK] Unable to set watch for target at path: '%s'", watchPath)
-			return err
-		} else {
-			z.config.Log.Info().Msgf("[ZK] Successfuly set data watch for: %s", watchPath)
+	// TODO: Fix reconnect issue for target connections
+	// if err := z.refreshZookeeperConfig(z.config.Log, targetChan, z.config); err != nil {
```
Comment (petrutlucian94) [inline gateway/loaders/zookeeper/zookeeper.go]: leftover?

**Review by petrutlucian94 — COMMENTED** [review]
just a few minor nits. apart from that, we'll need to fix the target reconnect workflow.

File: gateway/connections/zookeeper.go [hunk 4]
```diff
@@ -187,6 +187,64 @@ func (c *ZookeeperConnectionManager) handleTargetControlMsg(msg *TargetConnectio
...
 		}
 	}
+
+	if msg.ReconnectAll {
+		// targets := make(map[string]*targetpb.Target)
+		// targetNames := []string{}
+		for _, conn := range c.connections {
+			// c.config.Log.Info().Msgf("Reconnecting conn: ", conn)
+			if err := conn.reconnect(c.connLimit); err != nil {
+				c.config.Log.Error().Msgf("Recconect failed for connection: ", err)
```
Comment (petrutlucian94) [inline gateway/connections/zookeeper.go]: nit: s/Recconect/Reconnect. also, "Reconnect failed" is enough, "Reconnect failed for connection" sounds a bit awkward.

File: gateway/loaders/zookeeper/utils.go (hunk 2 above)
Comment (stefan-chivu) [reply gateway/loaders/zookeeper/utils.go:68]: the check is performed for the TLS config path, specifically /certificates/global

Comment (petrutlucian94) [reply gateway/loaders/zookeeper/utils.go:68]: ohh, my bad, I think I misread that

**Comment by petrutlucian94** [comment]
I think this is ready to go now.

**Pull request merged**